### PR TITLE
Add implementation of silencers

### LINKED ISSUE #57: Add list of ignore hosts for healing
Sometimes, hosts are under maintainance/development might be misevaluated as down and needed for healing.

Should have a method to avoid that case.

any thoughts @ntk148v?

### PR DESCRIPTION
**Pull request by vtdat**
* Silence model
* APIs for create/list/expire silence

Closes #57

### CODE REVIEW

File: pkg/autohealer/manager.go [hunk 1]
```diff
@@ -359,7 +362,7 @@ func (hm *Manager) getATEngine(key string) (model.ATEngine, error) {
 	return atengine, nil
 }
 
-// MakeKey creates key from id and instance
-func MakeKey(id string, instance string) string {
+// makeKey creates key from id and instance
+func makeKey(id string, instance string) string {
```
Comment (ntk148v) [inline pkg/autohealer/manager.go]: Same here, please re-use the utility function.

https://github.com/vCloud-DFTBA/faythe/blob/37374ff636b4d99c4b29a1c76f22b95cd09e5f48/pkg/common/utils.go#L108

File: pkg/autohealer/healer.go [hunk 2]
```diff
@@ -64,10 +73,14 @@ func newHealer(l log.Logger, data []byte, b metrics.Backend, atengine model.ATEn
 
 func (h *Healer) run(ctx context.Context, wg *sync.WaitGroup, nc chan map[string]string) {
 	interval, _ := time.ParseDuration(h.Interval)
+	sinterval, _ := time.ParseDuration(model.DefaultSilenceValidationInterval)
 	duration, _ := time.ParseDuration(h.Duration)
 	ticker := time.NewTicker(interval)
+	sticker := time.NewTicker(sinterval)
 	chans := make(map[string]*chan struct{})
 	whitelist := make(map[string]struct{})
+	h.watchs = h.etcdcli.Watch(ctx, common.Path(model.DefaultSilencePrefix, h.CloudID), etcdv3.WithPrefix())
```
Comment (ntk148v) [inline pkg/autohealer/healer.go]: IMHO, the variable should be `swatch` or `silenceWatch` or something like that. It's much easier to follow up.

File: pkg/autohealer/healer.go [hunk 3]
```diff
@@ -106,35 +136,46 @@ func (h *Healer) run(ctx context.Context, wg *sync.WaitGroup, nc chan map[string
...
-					level.Info(h.logger).Log("msg", fmt.Sprintf("Not processed because the number of instance needed healing > %d", len(rIs)))
+				// If number of metrics returned for a instance != EvaluationLevel
+				// Or if instances in whitelist, delete from list of Instances, not process it
+				for k, v := range rIs {
+					if _, ok := whitelist[k]; ok || v != h.EvaluationLevel {
+						delete(rIs, k)
+					}
+				}
+
+				// If no of instances > DefaultMaxNumberOfInstances, clear all goroutines
```
Comment (ntk148v) [inline pkg/autohealer/healer.go]: s/no/number/g

File: pkg/autohealer/healer.go [hunk 4]
```diff
@@ -106,35 +136,46 @@ func (h *Healer) run(ctx context.Context, wg *sync.WaitGroup, nc chan map[string
...
+				// Or number of instances + number of existing instances need to heal > DefaultMaxNumberOfInstances
+				if len(rIs) > model.DefaultMaxNumberOfInstances || len(rIs)+len(chans) > model.DefaultMaxNumberOfInstances {
+					level.Info(h.logger).Log("msg",
+						fmt.Sprintf("Not processed because the number of instance needed healing = %d > %d",
```
Comment (ntk148v) [inline pkg/autohealer/healer.go]: s/Not/not/g

File: pkg/autohealer/healer.go [hunk 5]
```diff
@@ -47,15 +50,20 @@ type Healer struct {
 	mtx        sync.RWMutex
 	state      model.State
 	terminated chan struct{}
+	silences   map[string]*model.Silence
+	etcdcli    *common.Etcd
```
Comment (ntk148v) [inline pkg/autohealer/healer.go]: Create an instance of Etcd wrapper at manager level, then pass a watch chan as `run` argument.

File: pkg/model/silence.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+type Silence struct {
+	ID           string         `json:"id"`
+	Name         string         `json:"name"`
+	Pattern      string         `json:"pattern"`
+	RegexPattern *regexp.Regexp `json:"-"`
+	TTL          string         `json:"ttl"`
+	Tags         []string       `json:"tags"`
+	Description  string         `json:"description"`
+	// TODO: Do we need `created_at` ?
```
Comment (ntk148v) [inline pkg/model/silence.go]: We do, just keep it. Users may want to know when the silence was created.

Remove the TODO.

File: pkg/autohealer/healer.go (hunk 5 above)
Comment (vtdat) [reply pkg/autohealer/healer.go]: the reason I pass etcdcli is to update silences at starting point. is there any ways I can achieve it without passing `etcdcli` to healer?

Comment (ntk148v) [reply pkg/autohealer/healer.go]: :thinking: How about:
* Opt1: pass watchchan and getresp to run?
* Opt2: Pass etcdcli instance to run?

Comment (vtdat) [reply pkg/autohealer/healer.go]: @ntk148v recheck pls

**Comment by vtdat** [comment]
> Please add usage guide and the documentation.

already there: https://github.com/vCloud-DFTBA/faythe/blob/improve-documentation/docs/autohealing.md

**Pull request merged**